src/core/tsi/alts/crypt/aes_gcm.cc

```diff
@@ -687,21 +691,32 @@ grpc_status_code gsec_aes_gcm_aead_crypter_create(const uint8_t* key,
   aes_gcm_crypter->crypter.vtable = &vtable;
   aes_gcm_crypter->nonce_length = nonce_length;
   aes_gcm_crypter->tag_length = tag_length;
+  aes_gcm_crypter->is_key_owner = copy_key;
   if (rekey) {
     aes_gcm_crypter->key_length = kKdfKeyLen;
     aes_gcm_crypter->rekey_data = static_cast<gsec_aes_gcm_aead_rekey_data*>(
         gpr_malloc(sizeof(gsec_aes_gcm_aead_rekey_data)));
-    memcpy(aes_gcm_crypter->rekey_data->nonce_mask, key + kKdfKeyLen,
-           kAesGcmNonceLength);
+    if (copy_key) {
+      aes_gcm_crypter->rekey_data->nonce_mask =
+          static_cast<uint8_t*>(gpr_malloc(kAesGcmNonceLength));
+      memcpy(aes_gcm_crypter->rekey_data->nonce_mask, key + kKdfKeyLen,
+             kAesGcmNonceLength);
+    } else {
+      aes_gcm_crypter->rekey_data->nonce_mask = key + kKdfKeyLen;
```

Comment:
What's the plan here for when we do the rekey? Are we going to do that in a follow-up? Or do we seamlessly mutate the existing key even when we don't own it?

Reply:
IIUC of https://github.com/grpc/grpc/blob/948fd6bc9c9c0837c736c32fd229a4d9e5033f26/src/core/tsi/alts/crypt/aes_gcm.cc#L256, the derived key is stored separately in its own stack. If we want to protect that, we would have to derive this key on top of the non-owning buffer. However, since this derivation needs the original key, it seems to me we have only two choices:

1. Introduce SecretData (or its equivalent) here to protect this local buffer.
2. Have the caller provide two key buffers, one for mutation, one for read.

Whichever we do, including nothing, I'd prefer to do it after this PR. Since this is the one we need to unblock internal change.

Let me know if I'm missing something obvious.

Reply:
Sounds good! Done.